### PR TITLE
update docs for ContainerStatus fields

### PR DESCRIPTION
**Pull request by SergeyKanzhelev**
#### What type of PR is this?

/kind cleanup
/kind documentation

/sig node

#### What this PR does / why we need it:

Fixing stale documentation and align docs in API and internal types files. 

Helps with: https://github.com/kubernetes/kubernetes/issues/115199

#### Special notes for your reviewer:

There seems to be a problem with `Started` field in API. This seems to be wrong:

```
	// +optional
	Started     *bool
```

I wonder if we can change it to optional and not to be a pointer. 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: api/openapi-spec/v3/api__v1_openapi.json [hunk 1]
```diff
@@ -1405,17 +1405,17 @@
         "description": "ContainerStatus contains details for the current status of this container.",
         "properties": {
           "containerID": {
-            "description": "Container's ID in the format '<type>://<container_id>'.",
+            "description": "ContainerID is the ID of the container in the format '<type>://<container_id>'.",
```
Comment (ruiwen-zhao) [inline api/openapi-spec/v3/api__v1_openapi.json:1408]: Just for my own education, does the description always need to start with the property name?

File: staging/src/k8s.io/api/core/v1/types.go [hunk 2]
```diff
@@ -2631,31 +2631,48 @@ type ContainerState struct {
...
 	ImageID string `json:"imageID" protobuf:"bytes,7,opt,name=imageID"`
-	// Container's ID in the format '<type>://<container_id>'.
+	// ContainerID is the ID of the container in the format '<type>://<container_id>'.
 	// +optional
 	ContainerID string `json:"containerID,omitempty" protobuf:"bytes,8,opt,name=containerID"`
-	// Specifies whether the container has passed its startup probe.
-	// Initialized as false, becomes true after startupProbe is considered successful.
-	// Resets to false when the container is restarted, or if kubelet loses state temporarily.
-	// Is always true when no startupProbe is defined.
+	// Started indicates whether the container has passed its startup probe.
```
Comment (thockin) [inline staging/src/k8s.io/api/core/v1/types.go:2670]: `startupProbe` comes after `postStart`, so if `startupProbe` is defined, then this is obviously after `postStart` has finished.

But what if `startupProbe` is not defined?  Do you want this field set *always* or only after `postStart` has finished?

File: api/openapi-spec/v3/api__v1_openapi.json (hunk 1 above)
Comment (SergeyKanzhelev) [reply api/openapi-spec/v3/api__v1_openapi.json:1408]: I don't think it is a requirement, at least not enforced one. I just normalized everything to look neat and be aligned with golang docs practices

File: staging/src/k8s.io/api/core/v1/types.go (hunk 2 above)
Comment (SergeyKanzhelev) [reply staging/src/k8s.io/api/core/v1/types.go:2670]: this is the current behavior - it only being set after `postStart` complete and I think this is very reasonable. Indicating that the container has fully started.

Unfortunate naming is a clash with `StartedAt` field which indicates that process has started. I mentioned it in https://github.com/kubernetes/kubernetes/issues/115553

File: staging/src/k8s.io/api/core/v1/types.go [hunk 3]
```diff
@@ -2631,31 +2631,48 @@ type ContainerState struct {
...
+	// Each container in a pod must have a unique name across all container types.
 	// Cannot be updated.
 	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
-	// Details about the container's current condition.
+	// State holds details about the container's current condition.
 	// +optional
 	State ContainerState `json:"state,omitempty" protobuf:"bytes,2,opt,name=state"`
-	// Details about the container's last termination condition.
+	// LastTerminationState holds the last termination state of the container to
+	// simplify debugging of container crashes and restarts. This field if not
```
Comment (mrunalp) [inline staging/src/k8s.io/api/core/v1/types.go:2642]: nit: simplify debugging of -> help debug container crashes and restarts.

Comment (mrunalp) [reply staging/src/k8s.io/api/core/v1/types.go:2642]: This field is not..

File: staging/src/k8s.io/api/core/v1/types.go [hunk 4]
```diff
@@ -2631,31 +2631,48 @@ type ContainerState struct {
...
+	// probes are specified, this field defaults to true once the container is
+	// fully started (see Started field).
+	//
+	// The value is typically used to determine whether a container is ready to
+	// accept traffic.
 	Ready bool `json:"ready" protobuf:"varint,4,opt,name=ready"`
-	// The number of times the container has been restarted.
+	// RestartCount holds the number of times the container has been restarted.
+	// Kubelet makes the best effort of always incrementing the value, but there
+	// are cases when the state may be lost due to node restarts, than the value
```
Comment (mrunalp) [inline staging/src/k8s.io/api/core/v1/types.go:2656]: restarts and then the value

File: staging/src/k8s.io/api/core/v1/types.go [hunk 5]
```diff
@@ -2631,31 +2631,48 @@ type ContainerState struct {
...
+	// may be reset to 0. The value is never negative.
 	RestartCount int32 `json:"restartCount" protobuf:"varint,5,opt,name=restartCount"`
-	// The image the container is running.
+	// Image is the image the container is running. The image may not match the
```
Comment (mrunalp) [inline staging/src/k8s.io/api/core/v1/types.go:2659]: image that the..

File: api/openapi-spec/v3/api__v1_openapi.json [hunk 6]
```diff
@@ -1405,17 +1405,17 @@
...
             "type": "string"
           },
           "image": {
             "default": "",
-            "description": "The image the container is running. More info: https://kubernetes.io/docs/concepts/containers/images.",
+            "description": "Image is the image the container is running. The image may not match the image used in the PodSpec, as it may have been resolved by the runtime. More info: https://kubernetes.io/docs/concepts/containers/images.",
```
Comment (mrunalp) [inline api/openapi-spec/v3/api__v1_openapi.json:1413]: needs same changes here.

File: api/openapi-spec/v3/api__v1_openapi.json [hunk 7]
```diff
@@ -1425,26 +1425,26 @@
               }
             ],
             "default": {},
-            "description": "Details about the container's last termination condition."
+            "description": "LastTerminationState holds the last termination state of the container to simplify debugging of container crashes and restarts. This field if not populated if container is still running and RestartCount is 0."
```
Comment (mrunalp) [inline api/openapi-spec/v3/api__v1_openapi.json:1428]: Same changes here too.

Comment (SergeyKanzhelev) [reply api/openapi-spec/v3/api__v1_openapi.json:1428]: `make update` failed me, rerunning! Thank you for catching

**Comment by liggitt** [comment]
assigning @thockin since he already did a pass on this last week

cc @smarterclayton who has been slogging through container lifecycle stuff for visibility

File: pkg/apis/core/types.go [hunk 8]
```diff
@@ -2412,24 +2412,52 @@ type ContainerState struct {
...
-	// Each container in a pod must have a unique name.
+	// Name is a DNS_LABEL representing the unique name of the container.
+	// Each container in a pod must have a unique name across all container types.
+	// Cannot be updated.
 	Name string
+	// State holds details about the container's current condition.
 	// +optional
 	State ContainerState
+	// LastTerminationState holds the last termination state of the container to
+	// help debug container crashes and restarts. This field if not
```
Comment (thockin) [inline pkg/apis/core/types.go:2425]: s/if/is/

File: pkg/apis/core/types.go [hunk 9]
```diff
@@ -2412,24 +2412,52 @@ type ContainerState struct {
...
+	// populated if container is still running and RestartCount is 0.
```
Comment (thockin) [inline pkg/apis/core/types.go:2426]: ...if the container...

File: pkg/apis/core/types.go [hunk 10]
```diff
@@ -2412,24 +2412,52 @@ type ContainerState struct {
...
 	// +optional
 	LastTerminationState ContainerState
-	// Ready specifies whether the container has passed its readiness check.
+	// Ready specifies whether the container has passed its readiness check. The
```
Comment (thockin) [inline pkg/apis/core/types.go:2429]: maybe s/has passed/is currently passing/ ?

File: pkg/apis/core/types.go [hunk 11]
```diff
@@ -2412,24 +2412,52 @@ type ContainerState struct {
...
+	// probes are specified, this field defaults to true once the container is
+	// fully started (see Started field).
+	//
+	// The value is typically used to determine whether a container is ready to
+	// accept traffic.
 	Ready bool
-	// Note that this is calculated from dead containers.  But those containers are subject to
-	// garbage collection.  This value will get capped at 5 by GC.
+	// RestartCount holds the number of times the container has been restarted.
+	// Kubelet makes the best effort of always incrementing the value, but there
```
Comment (thockin) [inline pkg/apis/core/types.go:2438]: s/makes the best effort of always incrementing/makes an effort to always increment/ - "best effort" is an adjective.  Thanks, English!?!

File: pkg/apis/core/types.go [hunk 12]
```diff
@@ -2412,24 +2412,52 @@ type ContainerState struct {
...
+	// are cases when the state may be lost due to node restarts and than the value
```
[the remark this answers is not in the document]
Comment (thockin) [reply pkg/apis/core/types.go:2439]: +1

File: pkg/apis/core/types.go [hunk 13]
```diff
@@ -2412,24 +2412,52 @@ type ContainerState struct {
...
+	// may be reset to 0. The value is never negative.
 	RestartCount int32
-	Image        string
-	ImageID      string
+	// Image that the container is running. The image may not match the
```
Comment (thockin) [inline pkg/apis/core/types.go:2442]: "Container image" ?

File: pkg/apis/core/types.go [hunk 14]
```diff
@@ -2412,24 +2412,52 @@ type ContainerState struct {
...
+	// image used in the PodSpec, as it may have been resolved by the runtime.
+	Image string
+	// ImageID is the image ID of the container's image. The image ID may not
+	// match the image ID of the image used in the PodSpec, as it may have been
+	// resolved by the runtime.
+	ImageID string
+	// ContainerID is the ID of the container in the format '<type>://<container_id>'.
```
Comment (thockin) [inline pkg/apis/core/types.go:2449]: any clues on what "type" means?

File: pkg/apis/core/types.go [hunk 15]
```diff
@@ -2412,24 +2412,52 @@ type ContainerState struct {
...
 	// +optional
 	ContainerID string
-	Started     *bool
+	// Started indicates whether the container has passed its startup probe.
+	// Initialized as false, becomes true after startupProbe is considered
+	// successful. Resets to false when the container is restarted, or if kubelet
+	// loses state temporarily. In both cases, startup probes will run again.
+	// Is always true when no startupProbe is defined and container is running and
+	// has passed the postStart lifecycle hook. The null value must be treated the
```
Comment (thockin) [inline pkg/apis/core/types.go:2457]: I'd mention posStart earlier, if it matters.  Like:

"""
Started indicates whether the container has finished its postStart hook and passed its startup probe.
"""

File: pkg/apis/core/types.go (hunk 13 above)
Comment (SergeyKanzhelev) [reply pkg/apis/core/types.go:2442]: I wanted to start all help text with the name of the field. Or you mean the second word image?

Comment (thockin) [reply pkg/apis/core/types.go:2442]: Usually these are of the form `<fieldname> <verb> the <description>`.  "Image is the name of container image that the container is running"

File: staging/src/k8s.io/api/core/v1/types.go [hunk 16]
```diff
@@ -2631,31 +2631,49 @@ type ContainerState struct {
...
+	// probes are specified, this field defaults to true once the container is
+	// fully started (see Started field).
+	//
+	// The value is typically used to determine whether a container is ready to
+	// accept traffic.
 	Ready bool `json:"ready" protobuf:"varint,4,opt,name=ready"`
-	// The number of times the container has been restarted.
+	// RestartCount holds the number of times the container has been restarted.
+	// Kubelet makes the best effort of always incrementing the value, but there
+	// are cases when the state may be lost due to node restarts and than the value
```
[the remark this answers is not in the document]
Comment (SergeyKanzhelev) [reply staging/src/k8s.io/api/core/v1/types.go]: done

File: pkg/apis/core/types.go (hunk 15 above)
Comment (SergeyKanzhelev) [reply pkg/apis/core/types.go:2460]: done

File: pkg/apis/core/types.go (hunk 13 above)
Comment (SergeyKanzhelev) [reply pkg/apis/core/types.go]: updated second word

File: staging/src/k8s.io/api/core/v1/types.go [hunk 17]
```diff
@@ -2663,31 +2663,52 @@ type ContainerState struct {
...
+	// accept traffic.
 	Ready bool `json:"ready" protobuf:"varint,4,opt,name=ready"`
-	// The number of times the container has been restarted.
+	// RestartCount holds the number of times the container has been restarted.
+	// Kubelet makes an effort to always increment the value, but there
+	// are cases when the state may be lost due to node restarts and then the value
+	// may be reset to 0. The value is never negative.
 	RestartCount int32 `json:"restartCount" protobuf:"varint,5,opt,name=restartCount"`
-	// The image the container is running.
+	// Image that the container is running. The container image may not match the
```
Comment (marquiz) [inline staging/src/k8s.io/api/core/v1/types.go:2691]: following @thockin's suggestion, adding a verb

```suggestion
	// Image is the name of container image that the container is running. The container image may not match the
```

**Review by marquiz — COMMENTED** [review]
/lgtm

**Comment by SergeyKanzhelev** [comment]
@thockin can you please approve? @mrunalp is not around and he only had a few nits and was ok with the change in general

**Review by thockin — COMMENTED** [review]
Thanks!

/lgtm
/approve

**Review by alexsio27444 — COMMENTED** [review]
changes

**Pull request merged**